py/gc.c

```diff
@@ -805,27 +805,59 @@ void gc_dump_info(void) {
         (uint)info.num_1block, (uint)info.num_2block, (uint)info.max_block, (uint)info.max_free);
 }
 
-void gc_dump_alloc_table(void) {
+#if MICROPY_PY_MICROPYTHON_MEM_INFO_FRAGMENTATION
+// Power-of-two size classes. i.e. class[0] is how many single-block
+// allocations are possible. class[5] is how many 32-block allocations.
+#define NUM_FRAGMENTATION_CLASSES (20)
+
+// Given a number of contiguous blocks, figure out how many allocations of
+// each size class would fit in that many blocks.
+STATIC void gc_update_fragmentation_stats(size_t n, size_t *frag_classes) {
+    for (size_t c = 1, i = 0; c < n && i < NUM_FRAGMENTATION_CLASSES; c <<= 1, ++i) {
+        frag_classes[i] += n / c;
+    }
+}
+#endif
+
+#if MICROPY_PY_MICROPYTHON_MEM_INFO_BLOCKS || MICROPY_PY_MICROPYTHON_MEM_INFO_FRAGMENTATION
+void gc_dump_alloc_table(bool print_table, bool print_fragmentation) {
     GC_ENTER();
+    #if MICROPY_PY_MICROPYTHON_MEM_INFO_BLOCKS
     static const size_t DUMP_BYTES_PER_LINE = 64;
+    #endif
     #if !EXTENSIVE_HEAP_PROFILING
     // When comparing heap output we don't want to print the starting
```

Comment:
Now it looks as if this comment is for `if (print_table)` but it's actually for `#if !EXTENSIVE_HEAP_PROFILING` right?

Reply:
That's right.
But moving it above the #if seems weird too?

Reply:
Probably yes. It's a bit out of scope for this commit but to make this completely clear I'd go with changing the comment itself e.g. `Skip this when EXTENSIVE_HEAP_PROFILING is on because...`

Reply:
Done